File: backend/app/models/task.py
from sqlalchemy import Column, String, Text, Integer, ForeignKey, Table, Boolean, DateTime
from sqlalchemy.orm import relationship
from ..database import Base
import uuid
from datetime import datetime


# Промежуточная таблица для связи многие-ко-многим между задачами и исполнителями
task_assignees = Table(
    "task_assignees",
    Base.metadata,
    Column("task_id", String, ForeignKey("tasks.id", ondelete="CASCADE"), primary_key=True),
    Column("user_id", String, ForeignKey("users.id", ondelete="CASCADE"), primary_key=True)
)


class Task(Base):
    __tablename__ = "tasks"

    id = Column(String, primary_key=True, index=True, default=lambda: str(uuid.uuid4()))
    title = Column(String, nullable=False)
    description = Column(Text, nullable=True)
    column_id = Column(String, ForeignKey("task_columns.id", ondelete="CASCADE"), nullable=False)
    project_id = Column(String, ForeignKey("projects.id", ondelete="CASCADE"), nullable=False)
    order = Column(Integer, default=0)
    estimate_hours = Column(Integer, nullable=True)
    due_date = Column(DateTime, nullable=True)
    reviewer_id = Column(String, ForeignKey("users.id", ondelete="SET NULL"), nullable=True)
    created_at = Column(DateTime, default=datetime.utcnow)
    updated_at = Column(DateTime, default=datetime.utcnow, onupdate=datetime.utcnow)
    
    # Relationships
    column = relationship("TaskColumn", back_populates="tasks")
    project = relationship("Project")
    comments = relationship("Comment", back_populates="task", cascade="all, delete-orphan")
    
    # Многие-ко-многим с пользователями (исполнители)
    assignees = relationship("User", secondary=task_assignees, backref="assigned_tasks")
    reviewer = relationship("User", foreign_keys=[reviewer_id])

File: newbackend/src/main/java/tech/nicorp/pm/tasks/domain/Task.java
package tech.nicorp.pm.tasks.domain;

import jakarta.persistence.*;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;
import tech.nicorp.pm.projects.domain.Project;
import tech.nicorp.pm.projects.domain.TaskColumn;
import tech.nicorp.pm.users.domain.User;

import java.time.OffsetDateTime;
import java.util.ArrayList;
import java.util.List;
import java.util.UUID;

@Entity
@Table(name = "tasks")
@Getter
@Setter
@NoArgsConstructor
public class Task {
    @Id
    @Column(columnDefinition = "uuid")
    private UUID id = UUID.randomUUID();

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "project_id", nullable = false)
    private Project project;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "column_id", nullable = false)
    private TaskColumn column;

    @Column(name = "title", nullable = false)
    private String title;

    @Column(name = "description")
    private String description;

    @Column(name = "order_index", nullable = false)
    private Integer orderIndex = 0;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "reviewer_id")
    private User reviewer;

    @Column(name = "due_date")
    private OffsetDateTime dueDate;

    @Column(name = "estimate_minutes")
    private Integer estimateMinutes;

    @Column(name = "created_at", nullable = false)
    private OffsetDateTime createdAt = OffsetDateTime.now();

    @Column(name = "updated_at", nullable = false)
    private OffsetDateTime updatedAt = OffsetDateTime.now();

    @ManyToMany(fetch = FetchType.LAZY)
    @JoinTable(name = "task_assignees",
            joinColumns = @JoinColumn(name = "task_id"),
            inverseJoinColumns = @JoinColumn(name = "user_id"))
    private List<User> assignees = new ArrayList<>();

    @Column(name = "repository_id", columnDefinition = "uuid")
    private UUID repositoryId;

    @Column(name = "repository_branch")
    private String repositoryBranch;
}


